### PR TITLE
Functional tests

### PR DESCRIPTION
**Pull request by rgolangh**
This PR add a functional test suite. Run it with `make test-functional`
The suite needs a working cluster and a `KUBECONFIG` env var, or passing --kubeconfig to the test execution.

Run using: `make test-functional`

The suite executes and tests:
- create a VM (centos at the moment)
- create a tenant cluster - installs vanilla upstream k8s in the VM
- deploy the CSI driver
- create storage class, PVC and a running pod that consumes a PV
- tears down

### CODE REVIEW

**Review by ydayagi — APPROVED** [review]

**Comment by rgolangh** [comment]
Missing - using the CSI images under the test run

File: deploy/020-autorization.yaml [hunk 1]
```diff
@@ -31,6 +31,9 @@ rules:
   - apiGroups: [""]
     resources: ["persistentvolumeclaims/status"]
     verbs: ["update", "patch"]
+  - apiGroups: [ "storage.k8s.io" ]
+    resources: [ "volumeattachments/status" ]
+    verbs: [ "get", "list", "watch", "update", "patch" ]
```
Comment (isaacdorfman) [inline deploy/020-autorization.yaml:36]: Are all the verbs necessary ?

File: deploy/020-autorization.yaml [hunk 2]
```diff
@@ -105,17 +108,6 @@ rules:
...
----
-
-apiVersion: rbac.authorization.k8s.io/v1
-kind: ClusterRole
-metadata:
-  name: openshift:csi-driver-controller-leader-election
-rules:
-  - apiGroups: [""]
-    resources: ["configmaps", "endpoints"]
-    verbs: ["get", "list", "watch", "update", "create", "delete"]
```
Comment (isaacdorfman) [inline deploy/020-autorization.yaml:118]: Are all the verbs necessary ?

File: tests/functional/functional_test.go [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+				if p.Status.Phase == v1.PodRunning {
+					return true, nil
+				}
+			default:
+				return false, nil
+			}
+
+			return false, nil
+		},
+	)
```
Comment (isaacdorfman) [inline tests/functional/functional_test.go:264]: I think that this part should be a separate function.

Comment (rgolangh) [reply tests/functional/functional_test.go:264]: you didn't say which part exactly but I guess the until with sync?

File: hack/build-tests.sh [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+export PATH=$PATH:$HOME/gopath/bin
+JOB_TYPE="${JOB_TYPE:-}"
+
+if [ "${JOB_TYPE}" == "travis" ]; then
+    go get -v -t ./...
+    go get github.com/mattn/goveralls
+    go get -v github.com/onsi/ginkgo/ginkgo
+    go get -v github.com/onsi/gomega
+    go get -u github.com/evanphx/json-patch
+    go mod vendor
```
[the remark this answers is not in the document]
Comment (rgolangh) [reply hack/build-tests.sh:14]: Image has its own weight (creating downloding pushing downloading again) . Tests are anyway compiled, and they must be compatible with the code they test. If we had an image then we could easily enter inconsistency between the test and the code (code too new, dated tests)

File: tests/functional/functional_test.go (hunk 3 above)
Comment (isaacdorfman) [reply tests/functional/functional_test.go:264]: Yes

**Pull request merged**